libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -98,97 +117,187 @@ m_nameDispenser(_context.dispenser)
 
 void StackToMemoryMover::operator()(FunctionDefinition& _functionDefinition)
 {
-	for (TypedName const& param: _functionDefinition.parameters + _functionDefinition.returnVariables)
-		if (m_memoryOffsetTracker(param.name))
-		{
-			// TODO: we cannot handle function parameters yet.
-			return;
-		}
+	// It is important to first visit the function body, so that it doesn't replace the memory inits for
+	// variable arguments we might generate below.
 	ASTModifier::operator()(_functionDefinition);
+
+	vector<Statement> memoryVariableInits;
+
+	// All function parameters with a memory slot are moved at the beginning of the function body.
+	for (TypedName const& param: _functionDefinition.parameters)
+		if (auto slot = m_memoryOffsetTracker(param.name))
+			memoryVariableInits += generateMemoryStore(
+				m_context.dialect,
+				param.debugData,
+				*slot,
+				Identifier{param.debugData, param.name}
+			);
+
+	// All memory return variables have to be initialized to zero in memory.
+	for (TypedName const& returnVariable: _functionDefinition.returnVariables)
+		if (auto slot = m_memoryOffsetTracker(returnVariable.name))
+			memoryVariableInits += generateMemoryStore(
+				m_context.dialect,
+				returnVariable.debugData,
+				*slot,
+				Literal{returnVariable.debugData, LiteralKind::Number, "0"_yulstring, {}}
+			);
+
+	// Special case of a function with a single return argument that needs to move to memory.
+	if (_functionDefinition.returnVariables.size() == 1 && m_memoryOffsetTracker(_functionDefinition.returnVariables.front().name))
```

Comment:
Does this mechanism duplicate parameters or return variables, i.e. can it be that the two functions have local variables with the same name?

Reply:
If yes, they have to be disambiguated.

Reply:
It did in fact duplicate parameters. I just pushed a quick fix for that together with a test case for it.
This didn't become apparent so far, since we intentionally don't run any steps after the StackToMemoryMover that rely on disambiguation (at least I hope that's the reason :-)) - but even if that's the case, it's indeed better to keep things disambiguated here.